File: Graph/200_Number_Of_Islands/200_Number_Of_Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS, COLUMNS = len(grid), len(grid[0])
        seen = set()
        output = 0
        
        def dfs(row, col):                
            seen.add((row, col))
            directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]
            
            for dr, dc in directions:
                new_dr = dr + row
                new_dc = dc + col
                if (new_dr in range(ROWS) and
                    new_dc in range(COLUMNS) and
                    grid[new_dr][new_dc] == "1" and
                    (new_dr, new_dc) not in seen):
                    dfs(new_dr, new_dc)
                
        
        for row in range(ROWS):
            for col in range(COLUMNS):
                if (row, col) not in seen and grid[row][col] == "1":
                    output += 1
                    dfs(row, col)
                    
        return output

File: Graph/200_Number_Of_Islands/200_Number_Of_Islands.go
func numIslands(grid [][]byte) int {
    output := 0
    
    for i, row := range grid {
        for j, _ := range row {
            if grid[i][j] == '1' {
                visit(grid, i, j)
                output += 1
            }
        }
    }
    
    return output
}

func visit(grid [][]byte, row int, col int) {
    ROWS, COLUMNS := len(grid), len(grid[0])
    
    if (row < 0 || row >= ROWS || col < 0 || col >= COLUMNS || grid[row][col] == '0') {
        return
    }
    
    grid[row][col] = '0'
    
    directions := [][]int{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
    
    for _, direction := range directions {
        new_dr := direction[0] + row
        new_dc := direction[1] + col

        visit(grid, new_dr, new_dc)
    }
}
